### ISSUE TITLE
`Softplus` forward and backward discrepancy

### ISSUE THREAD

**Issue by shmsong** [pytorch/pytorch#55587]
## 🐛 Bug

`Softplus` operator backward implementation different from doc and forward implementation on applying `threshold`.

Documentation defines:
`For numerical stability the implementation reverts to the linear function when input×β>threshold` 
[https://pytorch.org/docs/stable/generated/torch.nn.Softplus.html](https://pytorch.org/docs/stable/generated/torch.nn.Softplus.html)

This is consistent with the behavior of `Softplus` forward kernels. However on the backward computation,  the `input * beta > threshold` condition becomes `output * beta> threshold`. As shown in below code snippet from 

`master @ b39eeb07ed2f60a27b60ad32d172b58d31554282`

```cpp

// wrapper from ATen/native/Activation.cpp
Tensor softplus_backward(
    const Tensor& grad_output,
    const Tensor& self,
    const Scalar& beta,
    const Scalar& threshold,
    const Tensor& output) {
  Tensor grad_input;
  auto iter = TensorIterator::binary_op(grad_input, grad_output, output);
  softplus_backward_stub(iter.device_type(), iter, beta, threshold);
  return iter.output();
}

//CPU impl from ATen/native/cpu/Activation.cpp
cpu_kernel_vec(
        iter,
        [beta, threshold](scalar_t a, scalar_t b) -> scalar_t {
          scalar_t z = std::exp(b * beta);
          return (b * beta) > threshold ? a : a * (z - scalar_t(1.)) / z;
        },
        [beta_vec, one_vec, threshold_vec](Vec a, Vec b) -> Vec {
          const Vec z = (b * beta_vec).exp();
          return Vec::blendv(a * (z - one_vec) / z, a, (b * beta_vec) > threshold_vec);
        }
    );

// CUDA impl from ATen/native/cuda/Activation.cu
 AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.dtype(), "softplus_backward_cuda", [&]() {
    auto beta = beta_.to<scalar_t>();
    auto threshold = threshold_.to<scalar_t>();
    gpu_kernel(iter, [beta, threshold]GPU_LAMBDA(scalar_t a, scalar_t b) -> scalar_t {
      scalar_t z = std::exp(b * beta);
      return (b * beta) > threshold ? a : a * (z - scalar_t(1.)) / z;
    });
  });

```

This makes backward of `Softplus` op semantically mismatched with forward. Even though with the default threshold of 20 it may not have observable effect, it can break `gradcheck` when the threshold is lowered. 

## To Reproduce

Thanks @zasdfgbnm for the repro:

```cpp
import torch
model = torch.nn.Softplus(threshold=1).double()
input = torch.tensor(0.9, dtype=torch.double, requires_grad=True)
output = model(input)
print(input.item(), output.item())
torch.autograd.gradcheck(model, input)
```

`Threshold` is really a method of improving numerical stability and current implementation shouldn't have problem with high `threshold` values (default 20 should be ok). It might make sense not to expose `threshold` as a parameter, as it is not a hyper-parameter to be tuned for sure. Updating all the `softplus_backward` to use the input to compute the gradients sounds like the the most thorough way. 

This issue is intended to be used for some quick discussions. @zasdfgbnm , @mruberry 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @albanD @gqchen @pearu @nikitaved @soulitzer @mruberry

**Comment by mruberry** [comment]
Thank you for reporting this issue, @shmsong. The team agrees it's a high priority.
